### PR TITLE
ENH: scil_resample_tractogram

### PR DESCRIPTION
**Pull request by AntoineTheb**
Script to resample a tractogram to a specific number of streamlines (in the dumbest way possible). If the tractogram has more streamlines than needed, the tractogram is subsampled to the number specified. If the tractogram has less streamlines, the tractogram is "upsampled" with new streamlines by adding noise to existing streamlines. The "upsampled" tractogram can be smoothed afterwards. 

For example, here is a u-shaped, hard-to-track bundle on the DiSCo3 dataset, upsampled to 1000 streamlines:

sub-DiSCo3_23

Allows to compensate for over or under-represented bundles. Kinda did that on a whim in 5 minutes so let me know if this script is useless or wrong or already exists. 

```
usage: scil_resample_tractogram.py [-h]
                                   (--point_wise_std POINT_WISE_STD | --streamline_wise_std STREAMLINE_WISE_STD)
                                   [--gaussian SIGMA | --spline SIGMA NB_CTRL_POINT]
                                   [-e ERROR_RATE]
                                   [--keep_invalid_streamlines]
                                   [--reference REFERENCE] [-f] [-v]
                                   in_tractogram nb_streamlines out_tractogram

Script to resample a tractogram to a set number of streamlines.
Can be useful to build training sets for machine learning algorithms, to
upsample under-represented bundles or downsample over-represented bundles.

Works by either selecting a subset of streamlines or by generating new
streamlines by adding gaussian noise to existing ones.

Includes smoothing to compensate for the noisiness of new streamlines
generated by the process.

Example usage:
$ scil_resample_tractogram.py input.trk 1000 output.trk --point_wise_std 0.5 --spline 5 10 --keep_invalid_streamlines
$ scil_visualize_bundles.py output.trk --local_coloring --width=0.1

positional arguments:
  in_tractogram         Input tractography file.
  nb_streamlines        Number of streamlines to resample the tractogram to.
  out_tractogram        Output tractography file.

optional arguments:
  -h, --help            show this help message and exit
  --point_wise_std POINT_WISE_STD
                        Noise to add to existing streamlines'points to generate new ones.
  --streamline_wise_std STREAMLINE_WISE_STD
                        Noise to add to existing whole streamlines to generate new ones.
  --gaussian SIGMA      Sigma for smoothing. Use the value of surronding
                        X,Y,Z points on the streamline to blur the streamlines.
                        A good sigma choice would be around 5.
  --spline SIGMA NB_CTRL_POINT
                        Sigma for smoothing. Model each streamline as a spline.
                        A good sigma choice would be around 5 and control point around 10.
  -e ERROR_RATE         Maximum compression distance in mm after smoothing. [0.1]
  --keep_invalid_streamlines
                        Keep invalid newly generated streamlines that may go out of the bounding box.
  --reference REFERENCE
                        Reference anatomy for tck/vtk/fib/dpy file
                        support (.nii or .nii.gz).
  -f                    Force overwriting of the output files.
  -v                    If set, produces verbose output.
```

To test:

```bash
$ ./scripts/scil_resample_tractogram.py ~/braindata/databases/DiSCo/derivatives/bundles/output/sub-DiSCo3/sub-DiSCo3_23.trk 1000 sub-DiSCo3_23_resmapled.trk --point=0.5 -f --keep --gaussian 5 --seed 1337
$ scil_visualize_bundles.py sub-DiSCo3_23_resmapled.trk --local_coloring --width=0.1
```

The upsampling part could definitely be improved in the future to generate less noisy streamlines.

@arnaudbore @GuillaumeTh @mdesco

### CODE REVIEW

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by AntoineTheb** [comment]
I realized a few days ago that it would be less messy to add the same "noise" to every point in the streamline instead of different noise to every point. This would essentially translate the streamline and thus give less noisy new streamlines. 

Will do once I have a few spare minutes.

File: scripts/scil_resample_tractogram.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    log_level = logging.WARNING
+    if args.verbose:
+        log_level = logging.DEBUG
+    logging.basicConfig(level=log_level)
+
+    sft = load_tractogram_with_reference(parser, args, args.in_tractogram)
+    original_number = len(sft.streamlines)
+
+    if args.nb_streamlines > original_number:
+        sft = upsample_tractogram(sft, args.nb_streamlines, args.std)
```
Comment (frheault) [inline scripts/scil_resample_tractogram.py:66]: You could smooth directly in there. Adding a smoothing parameter and smoothing streamlines directly after the point-wise noise.
Then you could apply the streamlines-wise noise.

File: scilpy/utils/streamlines.py [hunk 2]
```diff
@@ -664,3 +664,28 @@ def cut_invalid_streamlines(sft):
...
 
     return new_sft, cutting_counter
+
+
+def downsample_tractogram(sft, nb):
+    """ Downsample a tractogram to a specified number of streamlines.
+    """
+    new_indices = np.random.choice(
+        range(len(sft.streamlines)), nb, replace=False)
+    return filter_tractogram_data(sft, new_indices)
```
Comment (frheault) [inline scilpy/utils/streamlines.py:674]: sft[new_indices] should work

File: scripts/scil_resample_tractogram.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+""" Dumb script to resample a tractogram to a set number of streamlines.
```
Comment (mdesco) [inline scripts/scil_resample_tractogram.py:4]: Remove "Dumb" nothing is dumb ;)

File: scripts/scil_resample_tractogram.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Might be useful to build training sets for machine learning algorithms, to
```
Comment (mdesco) [inline scripts/scil_resample_tractogram.py:5]: Can be useful

File: scripts/scil_resample_tractogram.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+upsample under-represented bundles or downsample over-represented bundles.
+
+Works by either selecting a subset of streamlines or by generating new
+streamlines by adding gaussian noise to existing ones.
+
+It is recommended to use "scil_smooth_streamlines" afterwards to clean the
+noisy new streamlines.
```
Comment (mdesco) [inline scripts/scil_resample_tractogram.py:12]: Please provide your example 3 line execution as example, i.e. resample, smooth, visualize as in the PR conversation. This is really helpful for the future user

**Review by mdesco — APPROVED** [review]
Vraiment cool! I love it.

File: scilpy/utils/streamlines.py [hunk 6]
```diff
@@ -664,3 +664,28 @@ def cut_invalid_streamlines(sft):
...
+
+def upsample_tractogram(sft, nb, std):
+    """ Generate new streamlines to add to a tractogram samping a gaussian
+    centered around multiple streamlines existing points.
+    """
+    nb_new = nb - len(sft.streamlines)
+    indices = np.random.choice(
+        len(sft.streamlines), nb_new)
+    new_streamlines = sft.streamlines.copy()
+    for s in sft.streamlines[indices]:
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/utils/streamlines.py:685]: Not sure how I could generate new streamlines using sft.streamlines._data tho. Will keep the loop.

**Comment by AntoineTheb** [comment]
@frheault @mdesco @arnaudbore Addressed comments in 24d3e838be0e64d593e85292acd3ac69203c3590 and a0ef426742e278536ca1b037d15d3e05990523e2. Ready for re-review.

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-485/2/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-485/3/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-485/4/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by GuillaumeTh** [comment]
I think this script to do much things. We already have a script that downsample streamlines: scil_get_subset_streamlines.py. However, I really like the script to generate streamlines by adding noise.

File: scilpy/utils/streamlines.py [hunk 7]
```diff
@@ -3,14 +3,15 @@
 import itertools
 from functools import reduce
 import logging
+import numpy as np
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/utils/streamlines.py]: Fixed

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scilpy/utils/streamlines.py [hunk 8]
```diff
@@ -664,3 +665,40 @@ def cut_invalid_streamlines(sft):
...
 
     return new_sft, cutting_counter
+
+
+def upsample_tractogram(
+    sft, nb, point_wise_std=None,
+    streamline_wise_std=None, gaussian=None, spline=None, seed=None
+):
+    """ Generate new streamlines to add to a tractogram samping a gaussian
+    centered around multiple streamlines existing points.
```
Comment (arnaudbore) [inline scilpy/utils/streamlines.py:675]: clean docstring

File: scripts/scil_resample_tractogram.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                            ' X,Y,Z points on the streamline to blur the'
+                            ' streamlines.\nA good sigma choice would be '
+                            'around 5.')
+    sub_p.add_argument('--spline', nargs=2, metavar=('SIGMA', 'NB_CTRL_POINT'),
+                       type=int,
+                       help='Sigma for smoothing. Model each streamline as a '
+                            'spline.\nA good sigma choice would be around 5 '
+                            'and control point around 10.')
+
+    p.add_argument('-e', dest='error_rate', type=float, default=0.1,
```
Comment (arnaudbore) [inline scripts/scil_resample_tractogram.py:68]: -e or --e  ?

File: scripts/scil_resample_tractogram.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            parser.error("one of the arguments --point_wise_std " +
+                         "--streamline_wise_std is required")
+        sft = upsample_tractogram(
+            sft, args.nb_streamlines,
+            args.point_wise_std, args.streamline_wise_std,
+            args.gaussian, args.spline, args.seed)
+    elif args.nb_streamlines < original_number:
+        sft = get_subset_streamlines(sft, args.nb_streamlines, args.seed)
+    streamlines = compress_streamlines(
+        sft.streamlines, args.error_rate)
```
Comment (arnaudbore) [inline scripts/scil_resample_tractogram.py:115]: I would not use compressing within this script.

File: scripts/scil_resample_tractogram.py (hunk 9 above)
Comment (arnaudbore) [reply scripts/scil_resample_tractogram.py:68]: Please remove the compressing.

**Review by arnaudbore — APPROVED** [review]
Last comments otherwise LGTM

File: scripts/scil_resample_tractogram.py (hunk 9 above)
Comment (AntoineTheb) [reply scripts/scil_resample_tractogram.py]: Donesies in df4d41a44f3f2bfd5fcf3c4899142cfeccbc5675

File: scripts/scil_resample_tractogram.py (hunk 10 above)
Comment (AntoineTheb) [reply scripts/scil_resample_tractogram.py]: Yeah me neither df4d41a44f3f2bfd5fcf3c4899142cfeccbc5675

File: scilpy/utils/streamlines.py (hunk 8 above)
Comment (AntoineTheb) [reply scilpy/utils/streamlines.py]: Cleaned up and reworded in df4d41a44f3f2bfd5fcf3c4899142cfeccbc5675

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scilpy/utils/streamlines.py [hunk 11]
```diff
@@ -664,3 +665,41 @@ def cut_invalid_streamlines(sft):
...
+
+
+def upsample_tractogram(
+    sft, nb, point_wise_std=None,
+    streamline_wise_std=None, gaussian=None, spline=None, seed=None
+):
+    """ Generate new streamlines by either adding gaussian noise around
+    streamlines' points, or by translating copies of existing streamlines
+    by a random amount.
+    """
```
Comment (arnaudbore) [inline scilpy/utils/streamlines.py:677]: clean docstring

File: scripts/tests/test_resample_tractogram.py [hunk 12]
```diff
@@ -12,14 +12,15 @@
...
 def test_execution_tracking(script_runner):
     os.chdir(os.path.expanduser(tmp_dir.name))
     in_tracto = os.path.join(get_home(), 'tracking',
-                             'union_shuffle.trk')
-    ret = script_runner.run('scil_get_subset_streamlines.py', in_tracto,
-                            '1000', 'union_shuffle_sub.trk')
+                             'union_shuffle_sub.trk')
+    ret = script_runner.run('scil_resample_tractogram.py', in_tracto,
+                            '1000', 'union_shuffle_sub_resampled.trk',
+                            '--point_wise_std', '0.5')
```
Comment (arnaudbore) [inline scripts/tests/test_resample_tractogram.py:25]: add a test using upsampling

File: scripts/scil_resample_tractogram.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    p.add_argument('in_tractogram',
+                   help='Input tractography file.')
+    p.add_argument('nb_streamlines', type=int,
+                   help='Number of streamlines to resample the tractogram to.')
+    p.add_argument('out_tractogram',
+                   help='Output tractography file.')
+    std_group = p.add_mutually_exclusive_group()
+    std_group.add_argument('--point_wise_std', type=float,
+                           help='Noise to add to existing streamlines\'' +
+                                'points to generate new ones.')
```
Comment (arnaudbore) [inline scripts/scil_resample_tractogram.py:49]:
```suggestion
                                ' points to generate new ones.')
```

**Review by arnaudbore — APPROVED** [review]

File: scilpy/utils/streamlines.py (hunk 11 above)
Comment (AntoineTheb) [reply scilpy/utils/streamlines.py:677]: Clean quoi ?

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scilpy/utils/streamlines.py (hunk 11 above)
Comment (GuillaumeTh) [reply scilpy/utils/streamlines.py:703]: I think it means to add a description for each parameters.

Comment (AntoineTheb) [reply scilpy/utils/streamlines.py:703]: Donesoes in 93351ff

File: scripts/tests/test_resample_tractogram.py (hunk 12 above)
Comment (AntoineTheb) [reply scripts/tests/test_resample_tractogram.py:34]: Tests were reworked in 93351ff

**Review by GuillaumeTh — APPROVED** [review]

**Pull request merged**